Feature/tags

NathanealV opened the pull request:
- [Added tag create, delete, list, and toggle commands](https://github.com/ib-ai/ib.py/commit/821d44af16e6ade536867d0f3e0fce09878f16d7)
- [Added tags to cached.py](https://github.com/ib-ai/ib.py/commit/588fedc749fc58a475518a60d6f8067d9eb6f9e9)
- [Changed RegexConverter to class for better typing](https://github.com/ib-ai/ib.py/commit/7215b79409e8da2399078bb1658a6224bb2ce9da)
- [Added disableReply and GuildReply tables](https://github.com/ib-ai/ib.py/commit/e720efa0ba9f651ca31d2fa6e146bb3755ec4035)

NathanealV committed:
Added tag create, delete, list, and toggle commands

NathanealV committed:
Added tags to `cached.py`

NathanealV committed:
Changed `RegexConverter` to class for better typing

NathanealV committed:
Added `disableReply` and `GuildReply` tables

ilikecubesnstuff commented on db/models.py:
```diff
@@ -87,6 +87,13 @@ class Meta():
     finished = fields.BooleanField(default=False)
     vote_ladder_id = fields.OneToOneField('models.GuildVoteLadder')
 
+class GuildReply(Model):
```
Something like `GuildSuppressReply` is a more intuitive name.

ilikecubesnstuff replied:
Actually, I think this can be combined with the already existing `GuildData` model.

ilikecubesnstuff commented on utils/converters.py:
It could be worth changing all the converters into classes for consistency's sake.

ilikecubesnstuff commented on db/models.py:
```diff
@@ -87,6 +87,13 @@ class Meta():
+        table = "reply"
+
+    reply_id = fields.BigIntField(pk=True, unique=True)
+    channels = ArrayField(element_type="bigint", null = True)
```
Keyword argument spacing.

ilikecubesnstuff commented on db/cached.py:
```diff
@@ -1,18 +1,28 @@
+
+@alru_cache
+async def get_all_tags() -> list[StaffTag]:
+    return await StaffTag.all()
```
These share a pattern - the code could be condensed with a factory function, e.g.
```py
def model_cache_factory(Model):
    @alru_cache
    async def model_cache() -> list[Model]:
        return await Model.all()
    return model_cache
```

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
-    
+        if len(trigger) > NAME_SIZE_LIMIT - 10: # 2 for backticks and 8 for [Disabled]
+            await ctx.send(f"The tag trigger is too long (should not exceed {NAME_SIZE_LIMIT - 2} characters).")
+            return
```
Inconsistency between the if statement and reply?

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
+            tag.output = output
+            await tag.save()
+        else:
+            await StaffTag.create(trigger=trigger, output=output)
```
Compress this with `update_or_create`.

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
+        tags = await get_all_tags()
+
+        if trigger:
+            trigger = (await RegexConverter().convert(ctx, trigger)).lower()
```
Use `RegexConverter` in the function signature as a type annotation.

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
+
+        if trigger:
+            trigger = (await RegexConverter().convert(ctx, trigger)).lower()
+            tags = [tag for tag in tags if trigger and trigger in tag.trigger.lower()]
```
The truthiness of `trigger` is already checked, `trigger in tag.trigger.lower()` should be sufficient.

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
+        disable_reply = message.channel.id in (await GuildReply.get_or_create(reply_id = message.guild.id))[0].channels
+
+        if disable_reply:
+            return
```
If you merge `GuildReply` into `GuildData`, this part could be cleaner:
```py
guild_data = await get_guild_data(guild_id = message.guild.id)
if message.channel.id in guild_data.suppressed_replies:
    return
```

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
+            if tag.disabled:
+                continue
+
+            if re.fullmatch(tag.trigger.lower(), message.content.lower(), re.IGNORECASE):
```
I suggest keeping the tag's trigger in its original case. Wait for another opinion, but I remember the lowercasing of triggers causing unexpected behaviors for some regular expressions with special characters (like `\W`, for example).

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
+        get_all_tags.cache_clear()
+
+        logger.debug(f"Deleted tag {trigger}.")
+        await ctx.send(f"Deleted tag `{trigger}`.")
```
Adding the tag's output in the log and reply would be handy.

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
+            tags = [tag for tag in tags if trigger and trigger in tag.trigger.lower()]
+
+        names = [
+            f"{'[Disabled]' if tag.disabled else ''} `{tag.trigger}`" for tag in tags]
```
This appears to leave a space at the start of non-disabled tags. Is an f-string like ```f"{'[Disabled] ' if tag.disabled else ''}`{tag.trigger}`"``` what is intended?

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
+
+    @commands.hybrid_group()
+    async def reply(self, ctx: commands.Context):
+        """
+        Commands for handling channels where certain bot replies (e.g., tags) are disabled.
+        """
+        await available_subcommands(ctx)
 
+    @reply.command()
+    async def list(self, ctx: commands.Context):
+        """
+        List of disabled reply channels.
+        """
+        guild_reply = (await GuildReply.get_or_create(reply_id = ctx.guild.id))[0]
+        channels = [f"<#{channel_id}>" for channel_id in guild_reply.channels] if guild_reply.channels else ["`None`"]
+
+        await ctx.send(f"Replies for the following channels are disabled: {', '.join(channels)}")
+
+    @reply.command()
+    async def toggle(self, ctx: commands.Context, channel: discord.TextChannel):
+        """
+        Toggle if bot replies are disabled for specified channel.
+        """
+        guild_reply = (await GuildReply.get_or_create(reply_id = ctx.guild.id))[0]
+
+        if not guild_reply.channels:
+            guild_reply.channels = []
+
+        disabled = channel.id in guild_reply.channels
+
+        if disabled:
+            guild_reply.channels.remove(channel.id)
+        else:
+            guild_reply.channels.append(channel.id)
+
+        await guild_reply.save()
+
+        logger.debug(
+            f"Channel {channel.id} replies are now {'disabled' if not disabled else 'enabled'}.")
+        await ctx.send(f"Replies for <#{channel.id}> are now {'disabled' if not disabled else 'enabled'}.")
```
Perhaps move these into `guilddata.py` if you merge `GuildReply` with `GuildData`.

NathanealV replied:
Can be moved when the moderation PR have been merged - ignoring for now.

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
+        """
+        guild_reply = (await GuildReply.get_or_create(reply_id = ctx.guild.id))[0]
+
+        if not guild_reply.channels:
+            guild_reply.channels = []
```
Redundant?

ilikecubesnstuff reviewed (changes requested):
Thank you for implementing the tags feature! My main feedback is about merging the new `GuildReply` model with the existing `GuilData` model. Apart from that, I have a bunch more small comments in various places.

NathanealV committed:
Changed converter functions to use the `commands.Converter` class for consistency.

NathanealV committed:
Removed `GuildReply` and merged into `GuildData` schema.

NathanealV committed:
Formatting for `updates.py` and `bot.py`.

NathanealV committed:
Cleaned up redundant code for caches.

NathanealV committed:
Refactored tag handling, eliminated redundant code, adjusted logging statements, and formatting.

NathanealV commented:
- Removed `.lower()` when checking for tag triggers
- Consistency with checking for tag trigger length
- Replaced redundant code with `update_or_create`
- Adjusted logging and success message for tag delete
- Adjusted `reply` commands for changes to DB schema
- Formatting

ilikecubesnstuff commented on cogs/tags.py:
```diff
@@ -55,34 +56,36 @@ async def tag(self, ctx: commands.Context):
-            await ctx.send(f"The tag trigger is too long (should not exceed {NAME_SIZE_LIMIT - 2} characters).")
+        if len(
+                trigger
+        ) > NAME_SIZE_LIMIT - 11:  # 2 for backticks and 9 for [Disabled] with trailing whitespace
```
Doesn't `[Disabled] ` take up 11 characters? So including the 2 backticks, this should total 13?

NathanealV committed:
Fixed tag trigger check value

NathanealV committed:
Updated error message to include new value

NathanealV committed:
Fix typo in `embed_dict` in `pagination.py`
